### PR TITLE
topology-updater:compute pod set fingerprint

### PR DESCRIPTION
**Pull request by jlojosnegros**
Add an option to compute the fingerprint of the current pod set on each node.
Report this new fingerprint using an attribute in new released `v1alpha2` NRT object.

needs: #1053 
address: #1048 
see: [this issue](https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/issues/30) for more info about new NRT api version

### CODE REVIEW

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (ffromani) [inline pkg/nfd-topology-updater/nfd-topology-updater.go:141]: I think we should add some preparatory refactoring letting `Scan()` here return a `struct` like `ScanResponse` which will hold `[]PodResources` initially and then also `map[string]string`

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (ffromani) [inline pkg/nfd-topology-updater/nfd-topology-updater.go:175]: we shoudl evaluate here to wrap the args in a `struct`

File: pkg/resourcemonitor/podresourcesscanner.go
Comment (ffromani) [inline pkg/resourcemonitor/podresourcesscanner.go:42]: also here we should consider to move args in a struct, but this can be deferred in a later PR

File: pkg/resourcemonitor/podresourcesscanner.go
Comment (ffromani) [inline pkg/resourcemonitor/podresourcesscanner.go:209]: should we fail badly or should we update without the annotation? If it's an annotation, it seems excessive to fail because of this. I'll think more about this.

File: pkg/resourcemonitor/podresourcesscanner_test.go
Comment (ffromani) [inline pkg/resourcemonitor/podresourcesscanner_test.go:539]: hard-earned lesson from RTE: we should document which pods (namespace/name) we expect which lead to this fingerprint. Or, if possible, compute it in the tests.

Comment (ffromani) [reply pkg/resourcemonitor/podresourcesscanner_test.go:539]: (here and also in all the instances below)

**Review by ffromani — COMMENTED** [review]
thanks for the PR! initial review.
Do we have plans to add integration/e2e tests?

File: pkg/resourcemonitor/podresourcesscanner.go
Comment (jlojosnegros) [reply pkg/resourcemonitor/podresourcesscanner.go:209]: Yeah ... it somehow seems too much to fail due to an annotation but in the end I went for the failure just because the user has to explicitly activate this feature and I slightly prefer to fail ... but I had the same doubt

File: pkg/resourcemonitor/podresourcesscanner_test.go
Comment (jlojosnegros) [reply pkg/resourcemonitor/podresourcesscanner_test.go:539]: I agree on the documentation part but how could we compute the result if we are testing the compute algorithm itself?

File: go.mod
Comment (ffromani) [inline go.mod:13]: you want to pull 0.1.2 here to use `podfingerprint.Attribute`

File: pkg/resourcemonitor/podresourcesscanner_test.go
Comment (ffromani) [reply pkg/resourcemonitor/podresourcesscanner_test.go:539]: the compute algorithm itself is encapsulated in the `podfingerprint` package, isn't it?
In order to make the tests debuggable is (much much) better to document/show the expectations in terms of pod running on nodes. How this set translate in which fingerprint value is the `podfingerprint` package job indeed.

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (jlojosnegros) [reply pkg/nfd-topology-updater/nfd-topology-updater.go]: I have changed it to use `ScanResponse` 
I could have created a new struct including all the params ( right now only ZoneList) but as this is a "private" method that is called only once it does not seems worthy.

File: pkg/resourcemonitor/podresourcesscanner_test.go
Comment (jlojosnegros) [reply pkg/resourcemonitor/podresourcesscanner_test.go]: I have changed to compute the expected fingerprint using the list of expected pods. 
Also I have documented that fingerprint only depends on namespace and name of the running pods in a node.

**Comment by ffromani** [comment]
> I think we should create a seperate PR that upgrades NRT api to `v1alpha2` and not bundle it in this one(you are missing changes to kustomize and helm deployment to install new CRDs).

good point! I'm fine with both approaches (and yes your suggestion is cleaner indeed)

**Comment by PiotrProkop** [comment]
> > I think we should create a seperate PR that upgrades NRT api to `v1alpha2` and not bundle it in this one(you are missing changes to kustomize and helm deployment to install new CRDs).
> 
> good point! I'm fine with both approaches (and yes your suggestion is cleaner indeed)

I just feel like we can easily miss important things we need to do with an upgrade like advertising Policies and Scope with Attributes 😄

**Comment by ffromani** [comment]
> > > I think we should create a seperate PR that upgrades NRT api to `v1alpha2` and not bundle it in this one(you are missing changes to kustomize and helm deployment to install new CRDs).
> > 
> > 
> > good point! I'm fine with both approaches (and yes your suggestion is cleaner indeed)
> 
> I just feel like we can easily miss important things we need to do with an upgrade like advertising Policies and Scope with Attributes smile

Yes, thinking about it you're right. Let's split the update to its own PR

**Comment by jlojosnegros** [comment]
Let's split this PR and create a new one only for NRT API update to `v1alpha2`
/hold

**Comment by jlojosnegros** [comment]
New PR created to update NRT API version: #1053

**Review by ffromani — COMMENTED** [review]
/lgtm
thanks @jlojosnegros for addressing the comments, nice work here

**Comment by jlojosnegros** [comment]
just a rebase....

**Comment by ArangoGutierrez** [comment]
> /lgtm

/Approve
/lgtm

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (marquiz) [inline pkg/nfd-topology-updater/nfd-topology-updater.go:203]: Shouldn't we check existing attributes and replace old attributes with the same name instead of blindly appending?

File: pkg/resourcemonitor/podresourcesscanner.go
Comment (marquiz) [inline pkg/resourcemonitor/podresourcesscanner.go:217]: nit: no string formatting is used here so either

```suggestion
			klog.Info("podFingerprint: " + status.Repr())
```

or 
```suggestion
			klog.Infof("podFingerprint: %s", status.Repr())
```

**Review by marquiz — COMMENTED** [review]
I happened to try to run the e2e tests but failed on two different clusters with

```
    Feb 15 11:52:06.023: INFO: podFingerprint attributte "nodeTopologyPodsFingerprint" not found attribute "nodeTopologyPodsFingerprint" not found
    Feb 15 11:52:15.519: FAIL: Timed out after 60.003s.
```

**Comment by jlojosnegros** [comment]
rebase after #1054 was merged.

File: pkg/resourcemonitor/podresourcesscanner.go
Comment (jlojosnegros) [reply pkg/resourcemonitor/podresourcesscanner.go]: Thanks for the comment and the suggestions.

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (jlojosnegros) [reply pkg/nfd-topology-updater/nfd-topology-updater.go]: Good point ...  I have just created a new function although it could be a better way to merge them..

**Comment by jlojosnegros** [comment]
Address comments.

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (jlojosnegros) [reply pkg/nfd-topology-updater/nfd-topology-updater.go]: > I happened to try to run the e2e tests but failed on two different clusters with
> 
> ```
>     Feb 15 11:52:06.023: INFO: podFingerprint attributte "nodeTopologyPodsFingerprint" not found attribute "nodeTopologyPodsFingerprint" not found
>     Feb 15 11:52:15.519: FAIL: Timed out after 60.003s.
> ```

it seems like the Attribute is not being created in the NRT object ... 
I have checked and nothing pretty obvious pop up, so I need to look a little deeper. 
I will take a look next monday

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (marquiz) [inline pkg/nfd-topology-updater/nfd-topology-updater.go:282]: I'm a bit concerned that with this any time any attribute is updated you basically get the list/slice re-shuffled in random order. Possibly ideally, we'd preserve the order of lhs with new attributes just appended.

Also, the current implementation is merely like `mergeAttributes()`. But should it be possible to delete attributes with this?

Any thoughts @ffromani and @PiotrProkop ?

Would be good to get this utility function into the NRT api...

Comment (marquiz) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:282]: > So the question is do we want to make it work for more than one attribute from the start or just update podfingerprint for now.

Good question – I'm open to either approach. But if we have a generic function like this it should do the "right thing"

Comment (ffromani) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:282]: my take is this code should only manage the attributes @PiotrProkop mentioned above, leaving everything else alone (= ignoring anything else). `nfd-topology-updater` should be the source of truth in the cluster for these attributes, so it should be allowed to add/update/overwrite/delete them, but just them.

Comment (marquiz) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:282]: @ffromani sounds legit to me 👍

Comment (marquiz) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:282]: I would say it is, when thinking about patching the object for example. But wdyt @ffromani @PiotrProkop ?

Comment (ffromani) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:282]: I'm not against keeping attributes ordered if we can help it, because this minimizes the changes. IOW, we should unnecessarily change objects or their fields. However well-behaving, robust clients should NOT rely on any ordering and treat the sequence as unordered.

Comment (jlojosnegros) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:282]: > my take is this code should only manage the attributes @PiotrProkop mentioned above, leaving everything else alone (= ignoring anything else). `nfd-topology-updater` should be the source of truth in the cluster for these attributes, so it should be allowed to add/update/overwrite/delete them, but just them.

Right now this function only add the `ScanResponse` attributes a newly created NRT or the the readed one. 
`ScanResponse` only manage `podfingerprint.Attribute.
Sorry, I think I do not fully understand what the new function is suppose to do ... could you please elaborate a little bit? Thanks

Comment (marquiz) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:282]: Clients should survive, I agree. But changing the order on every update breaks diffing, patching etc

Comment (ffromani) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:282]: > Clients should survive, I agree. But changing the order on every update breaks diffing, patching etc

I think we are in agreement then. `nfd-topology-updater` should preserve the existing order of attributes and be smart about add/update/delete (note: I think delete is very unlikely). This is also easy to unit-test

Comment (ffromani) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:282]: @marquiz I hear you about adding utilities to NRT API. I'll work on a PR, but I'm not sure I can make it quick enough to be used in this PR

Comment (PiotrProkop) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:282]: The simplest case is that this function iterates over all Attributes, looking for `podfingerprint.Attribute` if it's there check if checksum should be updated if so, update it in place. If it's not there append it to the end of the list.

Comment (jlojosnegros) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:282]: > The simplest case is that this function iterates over all Attributes, looking for `podfingerprint.Attribute` if it's there -> check if checksum should be updated if so, update it in place. If it's not there append it to the end of the list.

I can implement this approach here and change it to use @ffromani utility functions when available.

Comment (ffromani) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:282]: works for me

Comment (marquiz) [reply pkg/nfd-topology-updater/nfd-topology-updater.go:282]: Sounds good

Comment (jlojosnegros) [reply pkg/nfd-topology-updater/nfd-topology-updater.go]: Addressed. Let me know what you think

File: pkg/resourcemonitor/podresourcesscanner.go
Comment (marquiz) [inline pkg/resourcemonitor/podresourcesscanner.go:140]: Should we in this case remove the existing pod fingerprint attribute?

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (marquiz) [inline pkg/nfd-topology-updater/nfd-topology-updater.go:274]:
```suggestion
	for idx, info := range *attrList {
```

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (jlojosnegros) [reply pkg/nfd-topology-updater/nfd-topology-updater.go]: I have changed the moment when podfingerprint is computed. 
I was doing if AFTER the resources were filtered (in podresourcescanner.go:152) and if no resource pass the filter the Attribute was not added at all. 
Now it is computed using all the resources which seems better because podfingerprint should resemble node resource status

File: pkg/resourcemonitor/podresourcesscanner.go
Comment (ffromani) [reply pkg/resourcemonitor/podresourcesscanner.go:140]: > Should we in this case remove the existing pod fingerprint attribute?

I'm fine both ways. From the scheduler plugin perspective, clients consuming this attribute must tolerate either attribute disappearing (a previous update added, a later update removes) values and obsolete values.

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (PiotrProkop) [inline pkg/nfd-topology-updater/nfd-topology-updater.go:269]:
```suggestion
	if attrList == nil {
```
I know it is correct but it looks odd to have nil first.

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (PiotrProkop) [inline pkg/nfd-topology-updater/nfd-topology-updater.go:284]: can you add unit test for this function?

File: pkg/resourcemonitor/podresourcesscanner.go
Comment (jlojosnegros) [reply pkg/resourcemonitor/podresourcesscanner.go:140]: Here we are computing the ScanResponse. Podfingerprint attribute will not be added to the ScanResponse in case of an error.

Removing the existing attribute in the NTR should be done in the nfd-topology-updater's `updateNodeResourceTopology` function,  when updating the attribute list. 
That can be done but will go agains this (https://github.com/kubernetes-sigs/node-feature-discovery/pull/1049#discussion_r1109505690) and will require some changes in the `updateAttribute` function.
I understood that those changes will wait for the new utility functions in @ffromani 's PR

File: pkg/nfd-topology-updater/nfd-topology-updater.go
Comment (ffromani) [reply pkg/nfd-topology-updater/nfd-topology-updater.go]: In general I'd have squashed the fixing commit here. Codewise looks reasonnable.

File: pkg/nfd-topology-updater/nfd-topology-updater_test.go
Comment (ffromani) [inline pkg/nfd-topology-updater/nfd-topology-updater_test.go:52]: nit: this looks more like `updatedAttribute` (I'd not resubmit just for this)

File: pkg/nfd-topology-updater/nfd-topology-updater_test.go
Comment (ffromani) [inline pkg/nfd-topology-updater/nfd-topology-updater_test.go:101]: nit: I'd inline in the caller (I'd not resubmit just for this)

File: pkg/nfd-topology-updater/nfd-topology-updater_test.go
Comment (jlojosnegros) [reply pkg/nfd-topology-updater/nfd-topology-updater_test.go]: Done while squashing the fixing commit

File: pkg/nfd-topology-updater/nfd-topology-updater_test.go
Comment (jlojosnegros) [reply pkg/nfd-topology-updater/nfd-topology-updater_test.go]: Done while squashing the fixing commit

**Review by ffromani — COMMENTED** [review]
/lgtm

File: cmd/nfd-topology-updater/main.go
Comment (jlojosnegros) [reply cmd/nfd-topology-updater/main.go]: > Thanks for working on this @jlojosnegros. I think the code is basically ready to be merged (I had one nit about the cmdline flag description).
> 
> However, I probably forgot to ask this before, but we need to update the documentation (`docs/reference/topology-updater-commandline-reference.md`) to describe the new command line flag.
> 
> Also, could you add support for the new flag in the help chart:
> 
> * add smth like `topologyUpdater.podSetFingerprint` into `deployment/helm/node-feature-discovery/values.yaml`
> * use it in `deployment/helm/node-feature-discovery/templates/topologyupdater.yaml`
> * document the new helm parameter in `docs/deployment/helm.md`

Thanks for the comment. 
Sorry, I did not see this lack of documentation before.

File: deployment/helm/node-feature-discovery/templates/topologyupdater.yaml
Comment (ffromani) [inline deployment/helm/node-feature-discovery/templates/topologyupdater.yaml:58]: typo here

Comment (jlojosnegros) [reply deployment/helm/node-feature-discovery/templates/topologyupdater.yaml]: thanks

**Review by ffromani — COMMENTED** [review]
/lgtm

updates to docs and helm chart seem good

**Review by marquiz — APPROVED** [review]
Thanks @jlojosnegros for the quick update. Looks good to me now 👍 If something's missing let's fix it in subsequent PRs

**Pull request merged**
